tutor-contrib-multi-plugin/tutor_multi_k8s_plugin/plugin.py

```diff
@@ -6,27 +6,45 @@
 
 from .__about__ import __version__
 
-hooks.Filters.CONFIG_DEFAULTS.add_item(("MULTI_K8S_VERSION", __version__))
-# This plugin assumes you are using a central Traefik load balancer on your
-# cluster but if you want, you could use something else like
-# nginx + cert_manager, though that takes more effort to set up. In any case,
-# this plugin sets up standard k8s Ingress objects on the k8s cluster, so you
-# just need to have some working Ingress controller that can detect them and
-# route traffic to them with automatic HTTPS cert provisioning.
-# Use this MULTI_K8S_INGRESS_CLASS_NAME setting to specify the
-# 'ingressClassName' so that your preferred load balancer will see each
-# Open edX instance.
-hooks.Filters.CONFIG_DEFAULTS.add_item(("MULTI_K8S_INGRESS_CLASS_NAME", "tutor-multi-traefik"))
-hooks.Filters.CONFIG_OVERRIDES.add_items([
-    # Don't use Caddy as a per-instance external web proxy, but do still use it
-    # for per-instance load balancing. Eventually it would be nice to get rid of
-    # it completely as it's not needed in this setup, but that would break
-    # compatibility with existing Tutor k8s plugins that configure Caddy via
-    # patches.
-    ("ENABLE_WEB_PROXY", False),
-    # We are using HTTPS
-    ("ENABLE_HTTPS", True),
-])
+config = {
+    "defaults": {
+        "VERSION": __version__,
+
+        # This plugin assumes you are using either ingress-nginx or Traefik as an
+        # ingress controller to provide you with a central load balancer.
+        # The standard Ingress object uses annotations to trigger the generation
+        # of certificates when using ingress-nginx + cert-manager. See: https://cert-manager.io/docs/usage/ingress/#supported-annotations
+        # In this case the ingress class must include the 'nginx' substring.
+        # In the case of Traefik, there is no such constrain.
+        # provisioned by cert-manager.
+        "INGRESS_CLASS_NAME": "tutor-multi-nginx",
+        # Currently there is no easy way to autodiscover a list of public hosts of your
+        # Open edX installation to add to your ingress routes. This especially evident
+        # when installing additional plugins such as tutor-ecommerce or tutor-minio.
+        # The workaround is to manually add a list of hosts to be routed to the caddy
+        # instance.
```

Comment:
Yeah, this is a bit of an issue. I'd like to see a more automatic solution in the future.

In the future, the `Ingress` objects could be defined in each tutor plugin, e.g. https://github.com/overhangio/tutor-ecommerce/blob/master/tutorecommerce/patches/k8s-services instead of requiring this separate plugin. If they use an ingress class name like `tutor-multi-nginx`, they'll be ignored in the "normal" Tutor k8s environment (which has no ingress controller) but they'll work automatically in our helm chart with its central load balancer. So I think that would be a nice solution. Otherwise, what you have here seems like a good approach for now.

Thoughts?

CC @regisb

Reply:
I think that could work, or maybe a different kind of patch like `k8s-ingress` or `k8s-ingress-route`.

I think allowing some kind of code introspection between plugins can open interesting possibilities, but this is mostly a passing thought I had at the moment. I need to explore the plugin API more in depth.

Reply:
Could we leverage the [Caddy API](https://caddyserver.com/docs/api-tutorial)? On a local installation of Open edX this is what I'm getting with `curl localhost:2019/config/` inside the caddy container:

```
{
   "apps" : {
      "http" : {
         "servers" : {
            "srv0" : {
               "listen" : [
                  ":80"
               ],
               "logs" : {
                  "logger_names" : {
                     "apps.local.overhang.io" : "log4",
                     "discovery.local.overhang.io" : "log2",
                     "ecommerce.local.overhang.io" : "log3",
                     "local.overhang.io" : "log0",
                     "preview.local.overhang.io" : "log0",
                     "studio.local.overhang.io" : "log1"
                  }
               },
               "routes" : [
                  {
                     "handle" : [
                        {
                           "handler" : "subroute",
                           "routes" : [
                              {
                                 "handle" : [
                                    {
                                       "handler" : "request_body",
                                       "max_size" : 10000000
                                    },
                                    {
                                       "encodings" : {
                                          "gzip" : {}
                                       },
                                       "handler" : "encode",
                                       "prefer" : [
                                          "gzip"
                                       ]
                                    },
                                    {
                                       "handler" : "reverse_proxy",
                                       "headers" : {
                                          "request" : {
                                             "set" : {
                                                "X-Forwarded-Port" : [
                                                   "80"
                                                ]
                                             }
                                          }
                                       },
                                       "upstreams" : [
                                          {
                                             "dial" : "discovery:8000"
                                          }
                                       ]
                                    }
                                 ]
                              }
                           ]
                        }
                     ],
                     "match" : [
                        {
                           "host" : [
                              "discovery.local.overhang.io"
                           ]
                        }
                     ],
                     "terminal" : true
                  },
                  {
                     "handle" : [
                        {
                           "handler" : "subroute",
                           "routes" : [
                              {
                                 "handle" : [
                                    {
                                       "handler" : "request_body",
                                       "max_size" : 10000000
                                    },
                                    {
                                       "encodings" : {
                                          "gzip" : {}
                                       },
                                       "handler" : "encode",
                                       "prefer" : [
                                          "gzip"
                                       ]
                                    },
                                    {
                                       "handler" : "reverse_proxy",
                                       "headers" : {
                                          "request" : {
                                             "set" : {
                                                "X-Forwarded-Port" : [
                                                   "80"
                                                ]
                                             }
                                          }
                                       },
                                       "upstreams" : [
                                          {
                                             "dial" : "ecommerce:8000"
                                          }
                                       ]
                                    }
                                 ]
                              }
                           ]
                        }
                     ],
                     "match" : [
                        {
                           "host" : [
                              "ecommerce.local.overhang.io"
                           ]
                        }
                     ],
                     "terminal" : true
                  },
                  {
                     "handle" : [
                        {
                           "handler" : "subroute",
                           "routes" : [
                              {
                                 "handle" : [
                                    {
                                       "handler" : "request_body",
                                       "max_size" : 1000000
                                    }
                                 ],
                                 "match" : [
                                    {
                                       "path" : [
                                          "/api/profile_images/*/*/upload"
                                       ]
                                    }
                                 ]
                              },
                              {
                                 "handle" : [
                                    {
                                       "handler" : "request_body",
                                       "max_size" : 4000000
                                    }
                                 ]
                              },
                              {
                                 "group" : "group2",
                                 "handle" : [
                                    {
                                       "handler" : "rewrite",
                                       "uri" : "/theming/asset/images/favicon.ico"
                                    }
                                 ],
                                 "match" : [
                                    {
                                       "path_regexp" : {
                                          "pattern" : "^/favicon.ico$"
                                       }
                                    }
                                 ]
                              },
                              {
                                 "handle" : [
                                    {
                                       "encodings" : {
                                          "gzip" : {}
                                       },
                                       "handler" : "encode",
                                       "prefer" : [
                                          "gzip"
                                       ]
                                    },
                                    {
                                       "handler" : "reverse_proxy",
                                       "headers" : {
                                          "request" : {
                                             "set" : {
                                                "X-Forwarded-Port" : [
                                                   "80"
                                                ]
                                             }
                                          }
                                       },
                                       "upstreams" : [
                                          {
                                             "dial" : "lms:8000"
                                          }
                                       ]
                                    }
                                 ]
                              }
                           ]
                        }
                     ],
                     "match" : [
                        {
                           "host" : [
                              "local.overhang.io",
                              "preview.local.overhang.io"
                           ]
                        }
                     ],
                     "terminal" : true
                  },
                  {
                     "handle" : [
                        {
                           "handler" : "subroute",
                           "routes" : [
                              {
                                 "handle" : [
                                    {
                                       "handler" : "request_body",
                                       "max_size" : 250000000
                                    }
                                 ]
                              },
                              {
                                 "group" : "group3",
                                 "handle" : [
                                    {
                                       "handler" : "rewrite",
                                       "uri" : "/theming/asset/images/favicon.ico"
                                    }
                                 ],
                                 "match" : [
                                    {
                                       "path_regexp" : {
                                          "pattern" : "^/favicon.ico$"
                                       }
                                    }
                                 ]
                              },
                              {
                                 "handle" : [
                                    {
                                       "encodings" : {
                                          "gzip" : {}
                                       },
                                       "handler" : "encode",
                                       "prefer" : [
                                          "gzip"
                                       ]
                                    },
                                    {
                                       "handler" : "reverse_proxy",
                                       "headers" : {
                                          "request" : {
                                             "set" : {
                                                "X-Forwarded-Port" : [
                                                   "80"
                                                ]
                                             }
                                          }
                                       },
                                       "upstreams" : [
                                          {
                                             "dial" : "cms:8000"
                                          }
                                       ]
                                    }
                                 ]
                              }
                           ]
                        }
                     ],
                     "match" : [
                        {
                           "host" : [
                              "studio.local.overhang.io"
                           ]
                        }
                     ],
                     "terminal" : true
                  },
                  {
                     "handle" : [
                        {
                           "handler" : "subroute",
                           "routes" : [
                              {
                                 "handle" : [
                                    {
                                       "handler" : "request_body",
                                       "max_size" : 2000000
                                    },
                                    {
                                       "encodings" : {
                                          "gzip" : {}
                                       },
                                       "handler" : "encode",
                                       "prefer" : [
                                          "gzip"
                                       ]
                                    }
                                 ]
                              },
                              {
                                 "handle" : [
                                    {
                                       "handler" : "static_response",
                                       "status_code" : 204
                                    }
                                 ],
                                 "match" : [
                                    {
                                       "path" : [
                                          "/"
                                       ]
                                    }
                                 ]
                              },
                              {
                                 "handle" : [
                                    {
                                       "handler" : "reverse_proxy",
                                       "headers" : {
                                          "request" : {
                                             "set" : {
                                                "Host" : [
                                                   "local.overhang.io"
                                                ]
                                             }
                                          }
                                       },
                                       "upstreams" : [
                                          {
                                             "dial" : "lms:8000"
                                          }
                                       ]
                                    }
                                 ],
                                 "match" : [
                                    {
                                       "path" : [
                                          "/api/mfe_config/v1*"
                                       ]
                                    }
                                 ]
                              },
                              {
                                 "handle" : [
                                    {
                                       "handler" : "reverse_proxy",
                                       "headers" : {
                                          "request" : {
                                             "set" : {
                                                "X-Forwarded-Port" : [
                                                   "80"
                                                ]
                                             }
                                          }
                                       },
                                       "upstreams" : [
                                          {
                                             "dial" : "mfe:8002"
                                          }
                                       ]
                                    }
                                 ]
                              }
                           ]
                        }
                     ],
                     "match" : [
                        {
                           "host" : [
                              "apps.local.overhang.io"
                           ]
                        }
                     ],
                     "terminal" : true
                  }
               ]
            }
         }
      }
   },
   "logging" : {
      "logs" : {
         "default" : {
            "exclude" : [
               "http.log.access.log2",
               "http.log.access.log3",
               "http.log.access.log0",
               "http.log.access.log1",
               "http.log.access.log4"
            ]
         },
         "log0" : {
            "encoder" : {
               "fields" : {
                  "common_log" : {
                     "filter" : "delete"
                  },
                  "request>headers" : {
                     "filter" : "delete"
                  },
                  "resp_headers" : {
                     "filter" : "delete"
                  },
                  "tls" : {
                     "filter" : "delete"
                  }
               },
               "format" : "filter",
               "wrap" : {
                  "format" : "json"
               }
            },
            "include" : [
               "http.log.access.log0"
            ],
            "writer" : {
               "output" : "stdout"
            }
         },
         "log1" : {
            "encoder" : {
               "fields" : {
                  "common_log" : {
                     "filter" : "delete"
                  },
                  "request>headers" : {
                     "filter" : "delete"
                  },
                  "resp_headers" : {
                     "filter" : "delete"
                  },
                  "tls" : {
                     "filter" : "delete"
                  }
               },
               "format" : "filter",
               "wrap" : {
                  "format" : "json"
               }
            },
            "include" : [
               "http.log.access.log1"
            ],
            "writer" : {
               "output" : "stdout"
            }
         },
         "log2" : {
            "encoder" : {
               "fields" : {
                  "common_log" : {
                     "filter" : "delete"
                  },
                  "request>headers" : {
                     "filter" : "delete"
                  },
                  "resp_headers" : {
                     "filter" : "delete"
                  },
                  "tls" : {
                     "filter" : "delete"
                  }
               },
               "format" : "filter",
               "wrap" : {
                  "format" : "json"
               }
            },
            "include" : [
               "http.log.access.log2"
            ],
            "writer" : {
               "output" : "stdout"
            }
         },
         "log3" : {
            "encoder" : {
               "fields" : {
                  "common_log" : {
                     "filter" : "delete"
                  },
                  "request>headers" : {
                     "filter" : "delete"
                  },
                  "resp_headers" : {
                     "filter" : "delete"
                  },
                  "tls" : {
                     "filter" : "delete"
                  }
               },
               "format" : "filter",
               "wrap" : {
                  "format" : "json"
               }
            },
            "include" : [
               "http.log.access.log3"
            ],
            "writer" : {
               "output" : "stdout"
            }
         },
         "log4" : {
            "encoder" : {
               "fields" : {
                  "common_log" : {
                     "filter" : "delete"
                  },
                  "request>headers" : {
                     "filter" : "delete"
                  },
                  "resp_headers" : {
                     "filter" : "delete"
                  },
                  "tls" : {
                     "filter" : "delete"
                  }
               },
               "format" : "filter",
               "wrap" : {
                  "format" : "json"
               }
            },
            "include" : [
               "http.log.access.log4"
            ],
            "writer" : {
               "output" : "stdout"
            }
         }
      }
   }
}
```

Reply:
Thanks @regisb . It seems like the Caddy API has the data we need, but I'm not sure how it works with the lifecycle of Open edX with Tutor. The Caddy API won't return data until after the instance's Caddy container is already deployed and configured. But the k8s `Ingress` object is generated and deployed at the same time. In other words we can't query the Caddy API when Tutor is generating the config with `tutor config save`, right?

Reply:
Yeah, you really don't want to query the Caddy API at configuration time. If we really to do some dynamic introspection, then it has to happen either at init-time or in a dedicated [`do ...` command](https://docs.tutor.overhang.io/reference/api/hooks/consts.html#tutor.hooks.Filters.CLI_DO_COMMANDS).